cpp/src/arrow/array/util.cc

```diff
@@ -538,7 +538,7 @@ class NullArrayFactory {
       out_->buffers.resize(3);
       out_->buffers[2] = buffer_;
 
-      child_length = 1;
+      child_length = length_ > 0 ? 1 : 0;
```

Comment:
Function executor would call MakeNullArray with length zero when the input length is zero. The current implementation of MakeNullArray has a bug when the type is `list(dense_union(string))`, which is fixed by this change.

Reply:
Added a test that would fail with 
```
Failed 'list_union_nulls->ValidateFull()' failed with Invalid: List child array invalid: Invalid: Union child array #0 invalid: Invalid: Offsets buffer size (bytes): 4 isn't large enough for length: 1 and offset: 0
/home/jinshang/arrow/cpp/src/arrow/array/validate.cc:566  ValidateOffsets(type, values.size())
/home/jinshang/arrow/cpp/src/arrow/array/validate.cc:157  ValidateBinaryLike(type)
```
without this change. Comment is updated too.